### PR TITLE
Move/shrink schedule when changing the competition's date

### LINKED ISSUE #3360: Handle existing schedule when changing the competition's date
Currently if you do the following:
  - create a competition on day X
  - create a schedule for this competition
  - change the competition's day to Y

Then any change to the schedule in an existing room will be rejected because the activities for day X still exist, and we validate that all activities should be included in the competition's timeframe.

The easy and manual fix is to remove the rooms which have activities (which will consequently delete these activities), and recreate them.

Possible better automatic fix:
  - when the competition's dates change, delete all activities that are not in the timeframe
  - when the competition's dates change from start_date X to start_date Y, add Y-X days to all activities dates

Combining both would give a reasonable behavior: if you don't shrink the number of days of your competition, the schedule you originally created will be correctly moved to your new competition's date.
If you shrink the number of days (eg: from 3 to 2), 2 days of the original schedule will be kept and moved, whereas activities on the third will be removed.

### PR DESCRIPTION
**Pull request by viroulep**
Fixes #3360.

The behavior is the following:
  - if the number of days is the same (which I expect to be the most common scenario), everything is moved by (new_date-old_date).
  - if the number of days is not the same:
    - move activities on the old start date to the new start_date
    - move the others to the new end_date

It's not perfect when moving from a 4+ days competition to a 2 days competition, but I don't think we should bother with this extremely rare (if not non-existent) scenario.

### CODE REVIEW

File: WcaOnRails/spec/models/competition_spec.rb [hunk 1]
```diff
@@ -802,4 +802,48 @@
...
+    def change_and_check_activities(new_start_date, new_end_date)
+      on_first_day, on_last_day = all_activities.partition { |a| a.start_time.to_date == competition.start_date }
+      # the factory define one activity per day, the two lines below are
+      # basically safe guards against a future change to the competition's factory.
+      expect(on_first_day).not_to be_empty
+      expect(on_last_day).not_to be_empty
+      competition.update(start_date: new_start_date,
+                         end_date: new_end_date)
+      on_first_day.map(&:reload)
+      on_last_day.map(&:reload)
```
Comment (jonatanklosko) [inline WcaOnRails/spec/models/competition_spec.rb]: Perhaps `all_activities.each(&:reload)` would do the job, as all the arrays refer to the same objects.

File: WcaOnRails/app/models/competition.rb [hunk 2]
```diff
@@ -446,6 +446,38 @@ def update_foreign_keys
...
+    if start_date && end_date
+      # NOTE: when doing the change we don't need to care about the timezone, as we just "move" all the datetime the same way
+      if number_of_days >= old_number_of_days
+        competition_activities.each do |a|
+          a.move_by((start_date - old_start_date).to_i.days)
+        end
+      else
+        # NOTE: this is an arbitrary chosen policy when shrinking competition dates.
+        # move all activities on start_date to new "start_date"
+        # move all activities on days between ]start_date, end_date] to new "end_date"
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb]: Nit: `]start_date,` -> `[start_date,` ^^

File: WcaOnRails/spec/models/competition_spec.rb (hunk 1 above)
Comment (viroulep) [reply WcaOnRails/spec/models/competition_spec.rb]: Right thanks!

File: WcaOnRails/app/models/competition.rb (hunk 2 above)
Comment (viroulep) [reply WcaOnRails/app/models/competition.rb]: That's intentional: it's a notation to indicate that `start_date` is excluded from the interval :)

**Comment by viroulep** [comment]
>Two scenarios to keep in mind:
>  - one day is added to the competition (so only end_date gets changed by +1 day), currently we move all activities from day 2 to day 3, leaving day 2 empty, whereas one may expect the day 3 to be empty
>  - moving even from 3 days to 2 days means that we would make many activities overlap
>
>I don't have any fancier idea of dealing with it, just pointing out. Anything working is good for now (especially that the number of days is unlikely to be changed, as you said).

There is also the "add one day before" scenario, where something similar happens and activities are moved to day 1, leaving day 2 empty.
I couldn't find a "clean" way without introducing a lot of very specific cases, so I just went for something simple working for most use cases ^^

**Pull request merged**